.github/workflows/pypi-release.yml

```diff
@@ -2,157 +2,211 @@
-  #       with:
-  #         package: ${{ github.event.repository.name }}
-  #         version: ${{ inputs.tag }}
-  #         # test: ${{ inputs.testing }}  TODO: implement in action
-      
-  #     - name: Set output
-  #     # The above step will just use the latest version if the input version
-  #     # is not found.  So to validate the version we want to release exists
-  #     # we need to compare teh output version.
-  #       id: set_existence
-  #       run: |
-  #         if [[ ${{ steps.pypi_info.version }} != ${{ inputs.tag }} ]]
-  #         then
-  #           echo ${{ inputs.tag }} released to Pypi
-  #         else
-  #           echo ${{ inputs.tag }} FAILED to released to Pypi.  Manual intervention required.
-  #           exit 1
-  #         fi
+      - name: "[Notification] Package Version Already Live In Package Index. Skip Upload."
+        run: |
+          title="Package version already live in package index"
+          version=${{ inputs.version_number }}
+          package_name=${{ needs.sanitize-package-name.outputs.name }}
+          package_index="PyPI"
+          if [[ ${{ inputs.test_run }} == true ]]
+          then
+            package_index="Test PyPi"
+          fi
+          message="The version $version of package $package_name already lives in the $package_index. The upload to the package index will be skipped."
+          echo "::notice title=${{ env.NOTIFICATION_PREFIX }}: $title::$message"
+
+  test-pypi-release:
+    runs-on: ubuntu-latest
+    needs: [sanitize-package-name, check-package-exists-pypi]
+    if: >-
+      needs.check-package-exists-pypi.outputs.exists == 'false' &&
+      inputs.test_run == 'true'
+
+    environment: PypiTest
+
+    steps:
+      - name: "Download Build Artifact - ${{ inputs.version_number }}"
+        uses: actions/download-artifact@v3
+        with:
+          name: ${{ inputs.version_number }}
+          path: .
 
+      - name: "Publish ${{ needs.sanitize-package-name.outputs.name }} v.${{ inputs.version_number }} To Test PyPI"
+        uses: pypa/gh-action-pypi-publish@v1.6.4
+        with:
+          password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+          repository_url: https://test.pypi.org/legacy/
+
+  prod-pypi-release:
+    runs-on: ubuntu-latest
+    needs: [sanitize-package-name, check-package-exists-pypi]
+    if: >-
+      needs.check-package-exists-pypi.outputs.exists == 'false' &&
+      inputs.test_run == false
+
+    environment: PypiProd
+
+    steps:
+      - name: "Download Build Artifact - ${{ inputs.version_number }}"
+        uses: actions/download-artifact@v3
+        with:
+          name: ${{ inputs.version_number }}
+          path: .
+
+      - name: "Publish ${{ needs.sanitize-package-name.outputs.name }} v.${{ inputs.version_number }} To PyPI"
+        uses: pypa/gh-action-pypi-publish@v1.6.4
+        with:
+          password: ${{ secrets.PYPI_API_TOKEN }}
+
+  validate-package-available-pypi:
```

Comment:
Can there be any lag time between publishing a release and it being available?  Wondering if we could get any false negatives here.

Reply:
During the offline discussion, we agreed to introduce a retry option to the `py-package-info` GH action to cover this case.